src/actix/api_key.rs

```diff
@@ -0,0 +1,72 @@
+use std::future::{ready, Ready};
+
+use actix_web::body::{BoxBody, EitherBody};
+use actix_web::dev::{forward_ready, Service, ServiceRequest, ServiceResponse, Transform};
+use actix_web::{Error, HttpResponse};
+use futures_util::future::LocalBoxFuture;
+
+pub struct ApiKey {
+    api_key: String,
+}
+
+impl ApiKey {
+    pub fn new(api_key: &str) -> Self {
+        Self {
+            api_key: api_key.to_string(),
+        }
+    }
+}
+
+impl<S, B> Transform<S, ServiceRequest> for ApiKey
+where
+    S: Service<ServiceRequest, Response = ServiceResponse<EitherBody<B, BoxBody>>, Error = Error>,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Response = ServiceResponse<EitherBody<B, BoxBody>>;
+    type Error = Error;
+    type InitError = ();
+    type Transform = ApiKeyMiddleware<S>;
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        ready(Ok(ApiKeyMiddleware {
+            api_key: self.api_key.clone(),
+            service,
+        }))
+    }
+}
+
+pub struct ApiKeyMiddleware<S> {
+    api_key: String,
+    service: S,
+}
+
+impl<S, B> Service<ServiceRequest> for ApiKeyMiddleware<S>
+where
+    S: Service<ServiceRequest, Response = ServiceResponse<EitherBody<B, BoxBody>>, Error = Error>,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Response = ServiceResponse<EitherBody<B, BoxBody>>;
+    type Error = Error;
+    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    forward_ready!(service);
+
+    fn call(&self, req: ServiceRequest) -> Self::Future {
+        if let Some(key) = req.headers().get("api-key") {
+            if let Ok(key) = key.to_str() {
+                if self.api_key == key {
```

Comment:
We should probably do a constant time comparison here, because a simple comparison like this can leak our API key with a precision timing attack.

Maybe we can use: https://docs.rs/constant_time_eq/latest/constant_time_eq/fn.constant_time_eq.html

Reply:
From that page

> Compares two equal-sized byte strings in constant time.

and

> // Not equal-sized, so won't take constant time.

So we'd still potentially be leaking the key length.

---

I could pad both the key from the server and the key from the client to 1024 chars (using spaces or some constant character) before doing the comparison. At that point I should also check that the server key is less than equal to 1024 chars long.

Thoughts?

Reply:
@generall thoughts on adding the `constant_time_eq` crate as a dependancy?

Reply:
> So we'd still potentially be leaking the key length.

that is not that important, I think. Up to user to make it secure and not just `123`

Reply:
> @generall thoughts on adding the constant_time_eq crate as a dependancy?

I am ok with that

Reply:
> > So we'd still potentially be leaking the key length.
> 
> that is not that important, I think. Up to user to make it secure and not just `123`
> 

Agree. There'd still be enough entropy for a proper key.